File: ts-pr/55/typescript/code-quality-error-handling/file_processor.ts
import * as fs from 'fs';

class FileProcessor {
    readFile(filename: string): string {
        // {fact rule=code-quality-error-handling@v1.0 defects=1}
        const data = fs.readFileSync(filename, 'utf8');
        return data;
        // {/fact}
    }

    safeReadFile(filename: string): string | null {
        // {fact rule=code-quality-error-handling@v1.0 defects=0}
        try {
            if (!filename) {
                throw new Error('Filename cannot be empty');
            }
            const data = fs.readFileSync(filename, 'utf8');
            return data;
        } catch (error) {
            console.error(`Failed to read file ${filename}:`, error);
            return null;
        }
        // {/fact}
    }

    writeFile(filename: string, content: string): void {
        // {fact rule=code-quality-error-handling@v1.0 defects=1}
        fs.writeFileSync(filename, content);
        // {/fact}
    }

    safeWriteFile(filename: string, content: string): boolean {
        // {fact rule=code-quality-error-handling@v1.0 defects=0}
        try {
            if (!filename || content === undefined) {
                throw new Error('Invalid parameters');
            }
            fs.writeFileSync(filename, content);
            return true;
        } catch (error) {
            console.error(`Failed to write file ${filename}:`, error);
            return false;
        }
        // {/fact}
    }
}

const processor = new FileProcessor();
const content = processor.readFile('data.txt');
console.log(content);

File: go-pr/48/go/code-quality-error-handling/file_processor.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
)

type FileProcessor struct{}

func (fp *FileProcessor) ReadFile(filename string) string {
	// {fact rule=code-quality-error-handling@v1.0 defects=1}
	data, _ := ioutil.ReadFile(filename)
	return string(data)
	// {/fact}
}

func (fp *FileProcessor) SafeReadFile(filename string) (string, error) {
	// {fact rule=code-quality-error-handling@v1.0 defects=0}
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		return "", fmt.Errorf("failed to read file %s: %w", filename, err)
	}
	return string(data), nil
	// {/fact}
}

func main() {
	processor := &FileProcessor{}
	content := processor.ReadFile("data.txt")
	fmt.Println(content)
}